cirq/optimizers/controlled_gate_decomposer.py

```diff
@@ -0,0 +1,284 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List
+
+from cirq import linalg, ops
+import numpy as np
+
+
+def _unitary_power(U, p):
+    """Raises unitary matrix U to power p."""
+    assert linalg.is_unitary(U)
+    eig_vals, eig_vectors = np.linalg.eig(U)
+    Q = np.array(eig_vectors)
+    return Q @ np.diag(np.exp(p * 1j * np.angle(eig_vals))) @ Q.conj().T
+
+
+def _is_identity(M):
+    """Checks whether M is identity."""
+    return np.allclose(M, np.eye(M.shape[0]))
+
+
+def _ry(x):
+    return np.array([[np.cos(x / 2), np.sin(x / 2)],
+                     [-np.sin(x / 2), np.cos(x / 2)]])
+
+
+def _rz(x):
+    return np.diag(np.exp([0.5j * x, -0.5j * x]))
+
+
+def _flatten(x):
+    return sum(x, [])
+
+
+class ControlledGateDecomposer:
+    """Decomposes single-qubit multi-controlled gate.
+
+    Decomposition result is a sequence of single-qubit gates,
+    CNOT gates and (optionally) CCNOT gates.
+
+    === REFERENCE ===
+    Barenco, Bennett et al. Elementary gates for quantum computation. 1995.
+    https://arxiv.org/pdf/quant-ph/9503016.pdf
+    """
+
+    def __init__(self, allow_toffoli: bool = True):
+        """
+        Args:
+            allow_toffoli - whether Toffoli (CCNOT) gates are allowed in the
+               output. If `False`, output will consist only of single-qubit
+               gates and CNOT gates. Defaults to `True`.
+        """
+        self.allow_toffoli = allow_toffoli
+
+    def _ccnot_true(self, c0: 'cirq.Qid', c1: 'cirq.Qid', target: 'cirq.Qid'):
+        """Implements action of Toffoli gate."""
+        return [
+            ops.H.on(target),
+            ops.CNOT.on(c1, target),
+            ops.ZPowGate(exponent=-0.25).on(target),
+            ops.CNOT.on(c0, target),
+            ops.T.on(target),
+            ops.CNOT.on(c1, target),
+            ops.ZPowGate(exponent=-0.25).on(target),
+            ops.CNOT.on(c0, target),
+            ops.T.on(c1),
+            ops.T.on(target),
+            ops.H.on(target),
+            ops.CNOT.on(c0, c1),
+            ops.T.on(c0),
+            ops.ZPowGate(exponent=-0.25).on(c1),
+            ops.CNOT.on(c0, c1),
+        ]
+
+    def _ccnot_congruent(self, c0: 'cirq.Qid', c1: 'cirq.Qid',
+                         target: 'cirq.Qid'):
+        """Implements 'congruent' CCNOT.
+
+        This is almost like CCNOT, but multiplies 4th qubit by -1.
+        See lemma 6.2 in [1]."""
+        return [
+            ops.ry(-np.pi / 4).on(target),
+            ops.CNOT(c1, target),
+            ops.ry(-np.pi / 4).on(target),
+            ops.CNOT(c0, target),
+            ops.ry(np.pi / 4).on(target),
+            ops.CNOT(c1, target),
+            ops.ry(np.pi / 4).on(target),
+        ]
+
+    def _ccnot(self,
+               c0: 'cirq.Qid',
+               c1: 'cirq.Qid',
+               target: 'cirq.Qid',
+               congruent: bool = False):
```

Comment:
I would rename `congruent` to `allow_incorrect_phase`.

Reply:
I removed option to provide correct decomposition of Toffoli - instead I will just return cir.CCNOT and let user decompose it if they want.
So now we only have version with "congruent=True", so no need for this parameter.